### PR TITLE
Fix marker positions for unclosed paths

### PR DESCRIPTION
**Pull request by SilverCardioid**
TLDR: coordinates and angles in the array used to render markers get out of sync after an unclosed sub-path.

In path.py, `node.vertices` keeps track of the vertices of the path being parsed, and consists of alternating pairs of (x,y) coordinates and outgoing and incoming angles. The list is used by the draw_markers() function in the same file, which pops items off it two at a time to get the current point and a pair of angles.

When path() encounters a `z` ([closepath](https://www.w3.org/Graphics/SVG/1.1/paths.html#PathDataClosePathCommand)) command, `None` is appended to the list in place of a pair of angles to mark the end of a sub-path, with the following `M` or `m` (moveto) adding the starting point of a new sub-path. If an `M` or `m` is used without a preceding `z`, however, only the new starting point is appended, resulting in two consecutive pairs of coordinates in the array. This break of parity causes draw_markers() to interpret angles as coordinates and vice versa for the remainder of the path (hence drawing the markers for the bottom sub-path near the origin in the top left in the attached image). The changed file adds a `None` to the array if an `M` or `m` is not preceded by a `z`.

There are more discrepancies between CairoSVG's behaviour for markers and [the SVG specification](https://www.w3.org/Graphics/SVG/1.1/painting.html#Markers) visible in the image that I haven't looked into: start/end markers are placed at the ends of every sub-path instead of only the first and last nodes overall, and some of the angles are off. This is an easy fix for the positions of the markers, though. My encounter with the bug was not related to markers in the first place, but involved using every other element of `node.vertices` to get the vertices of a path from a string.

A similar problem might still occur if a `z` is followed by a command other than `M`/`m`, which, though valid SVG, I don't remember having ever seen in an SVG file.

The input SVG:
```xml
<?xml version="1.0" encoding="utf-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="600" height="600">
<defs>
<marker id="te" viewBox="0 0 10 10" refX="1" refY="5" markerUnits="strokeWidth" markerWidth="4" markerHeight="3" orient="auto"><path d="M0,0 10,5 0,10z"/></marker>
<marker id="tm" viewBox="0 0 10 10" refX="5" refY="5" markerUnits="strokeWidth" markerWidth="4" markerHeight="3" orient="auto"><path d="M0,5 5,0 10,5 5,10z"/></marker>
<marker id="ts" viewBox="0 0 10 10" refX="9" refY="5" markerUnits="strokeWidth" markerWidth="4" markerHeight="3" orient="auto"><path d="M10,0 0,5 10,10z"/></marker>
</defs>
<rect width="600" height="600" fill="#fff"/>
<path d="M100,100A300,300 0 0,1 300,300A200,200 0 0,1 200,200zM300,100A300,300 0 0,1 500,300A200,200 0 0,1 400,200M100,300A300,300 0 0,1 300,500A200,200 0 0,1 200,400" fill="none" stroke="#000" stroke-width="10" marker-start="url(#ts)" marker-mid="url(#tm)" marker-end="url(#te)"/>
</svg>
```

### CODE REVIEW

**Comment by liZe** [comment]
Hello!

Thanks **a lot** for this pull request.

> TLDR: coordinates and angles in the array used to render markers get out of sync after an unclosed sub-path.

TLDR: you’re right

> There are more discrepancies between CairoSVG's behaviour for markers and [the SVG specification](https://www.w3.org/Graphics/SVG/1.1/painting.html#Markers)

Of course. As you said the position is broken, but the angle is also broken (because we use the direction of the segment even when it’s a curve). I’ve also met a problem with scale lately.

Yes. We should rewrite everything. 😄

**Comment by SilverCardioid** [comment]
Thanks, and sorry for writing such a wall of text to require a TLDR :)

From the test image it doesn't seem like the marker angles are based on straight segments. The [section for elliptical arcs](https://github.com/Kozea/CairoSVG/blob/master/cairosvg/path.py#L159) in path() does some maths to determine the tangent angles from the calculated arc centre `(xc,yc)` and the vertices:
```python
# Find the drawing angles
angle1 = point_angle(xc, yc, 0, 0)
angle2 = point_angle(xc, yc, xe, ye)
# Store the tangent angles
node.vertices.append((-angle1, -angle2))
```
I'm not sure why the angles are negated here; I would think they should be `(angle1 ± pi/2, angle2 ∓ pi/2)` (with the signs depending on the `sweep` direction), but draw_markers() itself and some of the other path commands also do some kind of mirroring using `pi - angle`, and the scaling and rotation `A` uses to render non-circular arcs might complicate things.

**Comment by liZe** [comment]
> Thanks, and sorry for writing such a wall of text to require a TLDR :)

That was a perfect wall of text! 😍

> From the test image it doesn't seem like the marker angles are based on straight segments.

I don’t remember writing that, but I probably did :smile:… The drawing is definitely wrong though, we should review all these angles.
